### ISSUE TITLE
Ansatz parameters not getting updated for custom qRNN

### ISSUE THREAD

**Issue by Viggi999999** [qiskit-community/qiskit-machine-learning#670]
### Environment

- **Qiskit Machine Learning version**: '0.24.1'
- **Python version**: 3.9
- **Operating system**: 

### What is happening?

Ansatz parameters not getting updated in custom QRNN model using TorchConnector

### How can we reproduce the issue?

I'm trying to develop a Hybrid-qRNN model to predict a cosine wave. For this, I'm passing 7 values of x(t) and the output is based on the 7th value's output. However, when I call loss.backward() on this predicted output vs the actual y value, the ansatz weights do not get updated. I defined the ansatz based on this paper: https://arxiv.org/abs/2302.03244, and used EstimatorQNN to define the qlayer. I then used TorchConnector as I want the model to compute grad only w.r.t the 7th value (i.e. I define an loop from i=0:7 and only compute grad on the output AFTER exiting the loop).

Here's the code to reproduce my results:

```
#Create ansatz with this
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
from numpy import pi
from qiskit.circuit import Parameter
from qiskit.circuit import ParameterVector
def ansatz_ckt():
    qreg_q = QuantumRegister(6, 'q')
    creg_c = ClassicalRegister(1, 'c')
    params1 = ParameterVector("w",24)
    circuit = QuantumCircuit(qreg_q,creg_c)
    circuit.rx(params1[1], qreg_q[0])
    circuit.rx(params1[2], qreg_q[1])
    circuit.rx(params1[3], qreg_q[2])
    circuit.rx(params1[4], qreg_q[3])
    circuit.rx(params1[5], qreg_q[4])
    circuit.rx(params1[6], qreg_q[5])
    circuit.rz(params1[7], qreg_q[0])
    circuit.rz(params1[8], qreg_q[1])
    circuit.rz(params1[9], qreg_q[2])
    circuit.rz(params1[10], qreg_q[3])
    circuit.rz(params1[11], qreg_q[4])
    circuit.rz(params1[12], qreg_q[5])
    circuit.rx(params1[13], qreg_q[0])
    circuit.rx(params1[14], qreg_q[1])
    circuit.rx(params1[15], qreg_q[2])
    circuit.rx(params1[16], qreg_q[3])
    circuit.rx(params1[17], qreg_q[4])
    circuit.rx(params1[18], qreg_q[5])
    circuit.cnot(qreg_q[0],qreg_q[1])
    circuit.rz(params1[19],qreg_q[1])
    circuit.cnot(qreg_q[0],qreg_q[1])
    circuit.cnot(qreg_q[1],qreg_q[2])
    circuit.rz(params1[20],qreg_q[2])
    circuit.cnot(qreg_q[1],qreg_q[2])
    circuit.cnot(qreg_q[2],qreg_q[3])
    circuit.rz(params1[21],qreg_q[3])
    circuit.cnot(qreg_q[2],qreg_q[3])
    circuit.cnot(qreg_q[3],qreg_q[4])
    circuit.rz(params1[22],qreg_q[4])
    circuit.cnot(qreg_q[3],qreg_q[4])
    circuit.cnot(qreg_q[4],qreg_q[5])
    circuit.rz(params1[23],qreg_q[5])
    circuit.cnot(qreg_q[4],qreg_q[5])
    circuit.cnot(qreg_q[5],qreg_q[0])
    circuit.rz(params1[0],qreg_q[0])
    circuit.cnot(qreg_q[5],qreg_q[0])
    return circuit,params1
#circuit.measure(qreg_q[0], creg_c[0])
```

I then create the circuit here:
```
import torch
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit.circuit import ParameterVector
from torch.nn import MSELoss
from qiskit.circuit import Parameter

from qiskit_machine_learning.connectors import TorchConnector
from qiskit_machine_learning.neural_networks import EstimatorQNN, SamplerQNN
from qiskit.quantum_info import SparsePauliOp

def MKCircuit(numQubits):
    

    params_i = [Parameter("input1")]
    qreg_q = QuantumRegister(6, 'q')
    creg_c = ClassicalRegister(1, 'c')
    circuit_x = QuantumCircuit(qreg_q)

    circuit_x.ry(params_i[0], qreg_q[0])
    circuit_x.ry(params_i[0], qreg_q[1])
    circuit_x.ry(params_i[0], qreg_q[2])
    #fm = QuantumCircuit(QuantumRegister(numQubits), ClassicalRegister(1))
    #input_parameters = ParameterVector("theta", 2*numQubits)
    #performing the input data encoding as a RY and a RZ for
    #2 bits of data per qubit
    #for i in range(numQubits):
    #    fm.ry(input_parameters[i], i)
    #    fm.rz(input_parameters[numQubits+i], i)
    circuit_x.barrier()
    ansatz,weights = ansatz_ckt()
    #print(ansatz)
    #ansatz.measure(0,0)
    # the ansatz is that upon entangling each qubit to another input parameter
    # the circuit will train like a bounded 5D clock
    #weights = ParameterVector("w", numQubits)
    '''ansatz = QuantumCircuit(QuantumRegister(numQubits+1), ClassicalRegister(1))
    for i in range(numQubits-1):
        ansatz.crx(weights[i], i, i+1)
    #dont forget to close the clock and entangle the last qubit to the first qubit
    ansatz.crx(weights[-1], numQubits-1,0)

    # now if all works as calculated, the result of the
    # 5D clock projected to this ancilla
    # qubit should train to the binary classification of my label map
    ansatz.cx(numQubits-1 , numQubits)
    #Only measure the ancilla qubit
    ansatz.measure(numQubits, 0)'''

    # weights = ParameterVector("w", 4)

    qnn_qc = QuantumCircuit(QuantumRegister(numQubits), ClassicalRegister(1))
    qnn_qc.compose(circuit_x, inplace=True)
    qnn_qc.compose(ansatz, inplace=True)
    
    return qnn_qc, params_i, weights
```
This is the code I used to create the training data:

```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
def cos_preprocess():
    x_t = [i for i in range(0,100)]
    rad = 8/199
    y_t = []
    t_idk = []
    for i in x_t:
        t_dash = rad*i
        t_idk.append(t_dash)
        y_t.append(np.cos(np.pi*t_dash)/2)
    #x_t = np.array(x_t)
    #y_t = np.array(y_t)
    #x_t = np.reshape(x_t, newshape = (x_t.shape[0],1))
    #y_t = np.reshape(y_t, newshape = (y_t.shape[0],1))
    #for i in range(x_t.shape[0]):
    #    print(x_t[i][0])

    sc = MinMaxScaler(feature_range = (0, 1))

    X_train = []
    y_train = []
    x_t = np.array(x_t)
    x_t = x_t.reshape(-1,1)
    training_set_scaled = sc.fit_transform(x_t)
    for i in range(7, len(training_set_scaled)):
        X_train.append(training_set_scaled[i-7: i, 0])
        y_train.append(training_set_scaled[i, 0])
    X_train, y_train = np.array(X_train), np.array(y_train)
    X_train = np.reshape(X_train, newshape = (X_train.shape[0], X_train.shape[1], 1))
    return X_train,y_train
```
And finally, here is the code I used to train the model:
```
import matplotlib.pyplot as plt
def train_custom(qnn_qc, input_parameters, weights):

    #Does this measure the right qubit?
    op = SparsePauliOp.from_list([("IIIIIZ", 1)])
    qnn = EstimatorQNN(
    circuit=qnn_qc,
    input_params=input_parameters,
    weight_params=weights,
    observables = op
    
    )
    np.random.seed(42)
    estimator_qnn_weights = np.random.random(qnn.num_weights)
    #print(estimator_qnn_weights)
    model = TorchConnector(qnn)
    model.train()
    X_train,y_train = cos_preprocess()
    #print(X_train.shape)
    #features = torch.rand((20, 6))
    #labels = torch.randint(0, 2, (20, 2)) * 2. - 1

    optimizer = torch.optim.Adam(model.parameters(), lr=0.03)
    criterion = MSELoss()
    y_predicted = []
    epochs = 1
    total_loss = []
    hybrid_loss_list = []  # Store loss history

    for epoch in range(epochs):
        for i in range(len(X_train[:,0,:])):
            optimizer.zero_grad()
            if i%10==0:
                for name, params in model.named_parameters():
                    print(params)
            for j in range(0,7):
                
                
                output = model(torch.Tensor(X_train[i][j]))

            y_t = torch.tensor(output,requires_grad=True)
            y_idk = torch.Tensor(np.array(y_train[i]))
            y_idk = torch.reshape(y_idk, (-1,))
            #print(f"Predicted: {y_t}, Target: {y_idk}")
            #Save the losses to compare
            y_predicted.append(y_t)
            loss = criterion(y_t, y_idk)
            #print(f"Iteration: {i}, loss: {loss.item()}")
            loss.backward()
            optimizer.step()
            #print(loss)
            #Early stopping condition
            if len(total_loss)>0:
                if loss.item() > total_loss[-1]:
                    break
            total_loss.append(loss.item())
        
        hybrid_loss_list.append(sum(total_loss) / len(total_loss))
        print("Training [{:.0f}%]\tLoss: {:.4f}".format(100.0 * (epoch + 1) / epochs, hybrid_loss_list[-1]))
if __name__ == '__main__':
    
    circuit, input_params, weights = MKCircuit(6)
    print("Weights:",weights)
    train_custom(circuit, input_params, weights)
```
When I ran all of the above, my output was:
```
Weights: w, ['w[0]', 'w[1]', 'w[2]', 'w[3]', 'w[4]', 'w[5]', 'w[6]', 'w[7]', 'w[8]', 'w[9]', 'w[10]', 'w[11]', 'w[12]', 'w[13]', 'w[14]', 'w[15]', 'w[16]', 'w[17]', 'w[18]', 'w[19]', 'w[20]', 'w[21]', 'w[22]', 'w[23]']
Parameter containing:
tensor([ 0.1281,  0.0817,  0.1606,  0.3575, -0.1204, -0.4953, -0.2507, -0.2399,
        -0.1335, -0.6552, -0.7908,  0.8965, -0.3084,  0.9522,  0.8152, -0.8261,
         0.3550, -0.0571,  0.5820, -0.8997,  0.0799, -0.4646, -0.1589,  0.4438],
       requires_grad=True)
/tmp/ipykernel_15258/1790914909.py:46: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  y_t = torch.tensor(output,requires_grad=True)
Parameter containing:
tensor([ 0.1281,  0.0817,  0.1606,  0.3575, -0.1204, -0.4953, -0.2507, -0.2399,
        -0.1335, -0.6552, -0.7908,  0.8965, -0.3084,  0.9522,  0.8152, -0.8261,
         0.3550, -0.0571,  0.5820, -0.8997,  0.0799, -0.4646, -0.1589,  0.4438],
       requires_grad=True)
Parameter containing:
tensor([ 0.1281,  0.0817,  0.1606,  0.3575, -0.1204, -0.4953, -0.2507, -0.2399,
        -0.1335, -0.6552, -0.7908,  0.8965, -0.3084,  0.9522,  0.8152, -0.8261,
         0.3550, -0.0571,  0.5820, -0.8997,  0.0799, -0.4646, -0.1589,  0.4438],
       requires_grad=True)
Parameter containing:
tensor([ 0.1281,  0.0817,  0.1606,  0.3575, -0.1204, -0.4953, -0.2507, -0.2399,
        -0.1335, -0.6552, -0.7908,  0.8965, -0.3084,  0.9522,  0.8152, -0.8261,
         0.3550, -0.0571,  0.5820, -0.8997,  0.0799, -0.4646, -0.1589,  0.4438],
       requires_grad=True)
Training [100%] Loss: 0.0638
```
### What should happen?

The weight parameters from the ansatz do not change at all. The paper also mentions that I only need to measure the first qubit so I defined the SparsePauliOp.from_list parameter as ([("IIIIIZ", 1)]) but I'm not sure if this is the right way to do it. Any help on this would be greatly appreciated.

An alternate to implement this by creating a qRNN that takes 7 values of x(t) before outputting the y(t) needed for loss is also welcome!

Thanks in advance.

**Comment by Viggi999999** [comment]
Sorry if the formatting is unclear. I have also posted the question at https://stackoverflow.com/questions/76704968/ansatz-parameters-not-getting-updated-in-custom-qrnn-model-qiskit for clearer formatting of my code

**Comment by oscar-wallis** [comment]
Hi @Viggi999999, I quickly fixed your formatting and I am trying to figure out what your code is actually doing, I have quite a few questions before I can get down to what the problem is. I am gonna assume that the way you've built the circuit is correct. If you can answer my questions I can follow up on this. Also the code has a lot of commented lines which makes the purpose of the code unclear, if you could change this then that would be helpful too.

## Observable
If you just want to take the the z measurement of your first qubit then `SparsePauliOp.from_list parameter as ([("IIIIIZ", 1)])` will do exactly that, although you don't really need to specify having a classicalreg in your quantum circuit, its redundant.

## cos_preprocess()
The function isn't actually creating any cos related data, with the comments you have in now all its outputting is floats from 0 to 1 with step of 0.01 where every 8th number is the y_train ie:
```
X_train[0] = np.array([0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06)]
y_train[0] = np.array([0.07])
```
1. What is your input data?
2. What do you want your network to output/validate against?

Additionally, this line throws an error for me, I don't know what you are trying to do with this line either.
```
X_train = np.reshape(X_train, newshape = (X_train.shape[0], X_train.shape[1], 1))

- TypeError: only integer scalar arrays can be converted to a scalar index -
```
3. What's the purpose of the line above?

## train_custom()
So again this like 3D array style throws an error for me essentially due to the previous error.
```
for i in range(len(X_train[:,0,:])):
```
but if we forgive that and move onto this for loop:
```
for j in range(0, 7):
    output = model(torch.Tensor(X_train[i][j]))
```
All this is going to do is run the model 7 times, as you are using the same output name, you are re-writing the same variable and deleting the previous version. So, essentially, after that model has been run output is simply 'output = model(torch.Tensor(X_train[i][6])). The for loop is redundant and I am again not sure what you are trying to do. If you wanted to run the model with all 7 features of X_train, one at a time and your 'output' is then a tensor of the 7 individual runs then you could batch them and run model(torch.Tensor(X_train[i]), if you want to simply run the 7th X_train you could run model(torch.Tensor(X_train[i][6])). Either way, I think torch.Tensor is the wrong way to go about it, I think this isn't using torch syntax correctlty for these arrays, wherever you use torch.Tensor makes my device throw a warning,  better to use things like torch.from_numpy.
1. What do you want to input to your model here? 
2. What form do you want the output to take?

## Calculating the gradient
You said in this issue
'I then used TorchConnector as I want the model to compute grad only w.r.t the 7th value (i.e. I define an loop from i=0:7 and only compute grad on the output AFTER exiting the loop).' I want to clarify some things. 
1. When you say you want to compute the grad only w.r.t. the 7th value, what does this mean? 
If you are talking about the gradient of the weights/the process by which the weights are updated, this is done wrt ALL weights, as this is a variational quantum circuit and that is what they do internally for all specified weights. If instead you mean to calculate the loss of the network which is then later used to calculate a gradient / weight update, then:
2. Why run the first 6 numbers through the circuit if only the final input matters?  What instead I think you could mean 
3. Where is this done outside the loop? What loop? The training loop?

## Why aren't the weights updating?
I think to be honest there is a fair few bugs in your code, but there is a way you can get the weights to update. To make the weights actually update for the code you have here all you had to do is feed output directly into the loss instead of creating y_t e.g. change the loss line.
```
loss = criterion(output, y_idk)
```
It is worth mentioning something here though, even without the weights updating your loss decreases during an epoch. This is simply because your output and y_train become closer together simply due to the data leading to a more favourable answer. I honestly don't know way but it essentially made your network look like it was learning when really it wasn't. A good rule of thumb is to shuffle data, it stops overfitting and provides a more neutral training style, stops this kinda thing slipping you up.

## Where to go from here
I must admit, I have not read the paper you included a link to. If you can answer some of these questions and wanna give bug fixing this a go, I will put in that addtional effort and we can figure this out. I do appreciate this issue is almost a year old though -  I don't expect you want to pick this back up and I am sorry no one got to you sooner. 

## TLDR
Change the loss statement to `loss = criterion(output, y_idk)`.
